src/web/app/pages-instructor/instructor-course-edit-page/instructor-course-edit-page.component.ts

```diff
@@ -683,4 +692,146 @@ export class InstructorCourseEditPageComponent implements OnInit {
+            isArchived: true,
+            instructorCandidates: [],
+            instructorCandidatesSortBy: SortBy.NONE,
+            instructorCandidatesSortOrder: SortOrder.ASC,
+            hasInstructorsLoaded: false,
+            isTabExpanded: false,
+            hasLoadingFailed: false,
+          };
+          courseTabModels.push(model);
+        }
+      });
+    }, (err: ErrorMessageOutput) => {
+      this.isCopyingInstructor = false;
+      this.statusMessageService.showErrorToast(err.error.message);
+    }, () => {
+      const modalRef: NgbModalRef = this.ngbModal.open(CopyInstructorsFromOtherCoursesModalComponent);
+      modalRef.componentInstance.courses = courseTabModels;
+
+      modalRef.dismissed.subscribe(() => { this.isCopyingInstructor = false; });
+
+      modalRef.componentInstance.copyClickedEvent.subscribe((instructors: Instructor[]) => {
+        this.verifyInstructorsToCopy(instructors).subscribe((hasCheckPassed: boolean) => {
+          if (!hasCheckPassed) {
+            modalRef.componentInstance.isCopyingSelectedInstructors = false;
+            return;
+          }
+
+          this.addNewInstructors(instructors, modalRef);
+        });
+      });
+    });
+  }
+
+  /**
+   * Adds new instructors.
+   */
+  addNewInstructors(instructors: Instructor[], modalRef: NgbModalRef): void {
+    of(...instructors).pipe(
+      concatMap((instructor: Instructor) => {
+        return this.instructorService.createInstructor({
+          courseId: this.courseId,
+          requestBody: {
+            name: instructor.name,
+            email: instructor.email,
+            role: instructor.role!,
+            displayName: instructor.displayedToStudentsAs,
+            isDisplayedToStudent: instructor.isDisplayedToStudents!,
+          },
+        });
+      }),
+      // always close the modal after it enters the last step no matter adding succeeds or fails
+      finalize(() => {
+        this.isCopyingInstructor = false;
+        modalRef.componentInstance.isCopyingSelectedInstructors = false;
+        modalRef.close();
+      }),
+    ).subscribe((newInstructor: Instructor) => {
+      const newDetailPanels: InstructorEditPanelDetail = {
+        originalInstructor: { ...newInstructor },
+        originalPanel: this.getInstructorEditPanelModel(newInstructor),
+        editPanel: this.getInstructorEditPanelModel(newInstructor),
+      };
+      newDetailPanels.editPanel.permission = this.newInstructorPanel.permission;
+      newDetailPanels.originalPanel = JSON.parse(JSON.stringify(newDetailPanels.editPanel));
+
+      this.instructorDetailPanels.push(newDetailPanels);
+    }, (err: ErrorMessageOutput) => {
+      this.statusMessageService.showErrorToast(err.error.message);
+    }, () => {
+      this.statusMessageService.showSuccessToast(`The selected instructor(s) have been added successfully.
+      An email containing how to 'join' this course will be sent to them in a few minutes.`);
```

Comment:
Any reason for not putting this in the `next` block?

Reply:
If you're talking about the `complete` block, probably it's because of the usage of `concatMap`. If the toast is put at the end of the `next` block, it's shown immediately after the first instructor is created. On the contrast, if the toast is put in the `complete` block, it's shown after all instructors are added. I'm not very sure why though, so let me know if you know more implementation details.